### ISSUE TITLE
`unnecessary_parenthesis` false positive on (nullable) spread of an as-expression

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/linter#3816]
**Describe the issue**
False positive for detecting a ternary on lint `non_bool_condition` operation.

**To Reproduce**
```dart
[...?jsonDecode(str) as List?]
```

**Expected behavior**
No need to use parenthesis because it triggers `unnecessary_parenthesis` if solved doing this: 
```dart
[...?(jsonDecode(str) as List?)]
```

**Comment by srawlins** [comment]
The second example shows a false positive for unnecessary_parenthesis.

**Comment by lrhn** [comment]
That looks like a parser bug.
```dart
dynamic x = [0];
var l1 = [...? x as List<int>?]; // Fails to compile
var l2 = [...? x as List<int>?,]; // Perfectly fine.
```

The grammar says that what follows `...?` is an `expression`, so the `l1` line should work too.

> ```latex
> <spreadElement> ::= (`...' | `...?') <expression>
> ```

**Comment by oprypin** [comment]
The parser bug is fixed meanwhile. This no longer counts as a false positive, so this issue could just be closed.

But yes, perhaps let's come to some decision regarding spreads. I agree that `[...?(anything)]` should never trigger this lint. However, I also think that `[...(anything)]` also shouldn't trigger it. I've been mass-applying the lint and this is the biggest thing that stood out to me (other than the prior issues I've filed).

**Comment by srawlins** [comment]
Fixed by https://dart-review.googlesource.com/c/sdk/+/360400
